File: assignment2/RD_parser.py
import sys

EOF = -1
LETTER = 0
DIGIT = 1
UNKNOWN = 99

INT_LIT = 10
ADD_OP = 21 # '+'
SUB_OP = 22 # '-'
MULT_OP = 23 # '*'
DIV_OP = 24 # '/'
LEFT_PAREN = 25 # '('
RIGHT_PAREN = 26 # ')'

BUF_MAX = 1024

lexeme = [0 for i in range(BUF_MAX)]
cur_lexeme = [0 for i in range(BUF_MAX)]
nextChar = 0
nextToken = token = 0
charClass = 0
lexLen = 0
leftParenCnt = rightParenCnt = 0
checkCharCnt = 0

def expr():
    global nextToken
    left = term()
    rslt = left
    while nextToken == ADD_OP  or nextToken == SUB_OP:
        token = nextToken
        lex()
        right = term()
        if token == ADD_OP:
            rslt = add(left, right)
            left = rslt
        elif token == SUB_OP:
            rslt = sub(left, right)
            left = rslt
    return rslt    
    
def term():
    global nextToken
    left  = factor()
    rslt = left
    while nextToken == MULT_OP or nextToken == DIV_OP:
        token = nextToken
        lex()
        right = factor()
        if token == MULT_OP:
            rslt = mult(left, right)
            left = rslt
        elif token == DIV_OP:
            rslt = divd(left, right)
            left = rslt
    return rslt

def factor():
    global nextToken
    ifUnary = 0
    if nextToken == SUB_OP:
        ifUnary = 1
        lex()
        
    if nextToken == LEFT_PAREN:
        lex()
        exprVal = expr()
        if nextToken == RIGHT_PAREN:
            lex()
        else:
            print("Syntax error!!")
            exit()
        
        if ifUnary == 1:
            return -exprVal
        else:
            return exprVal   
    elif nextToken == INT_LIT:
        if ifUnary == 1:
            return -number()
        else:
            return number()
    else:
        print("Syntax error!!")
        exit()
        
def number():
    return digit()

def digit():
    global cur_lexeme, lexeme
    cur_lexeme.clear()
    i=0
    while lexeme[i] != '\0':
        cur_lexeme.append(lexeme[i])
        i += 1

    lex()
    return float("".join(cur_lexeme))

def lex():
    global lexLen, nextToken, lexeme, charClass, nextChar
    lexLen = 0
    getNonBlank()
    if charClass == LETTER:
        print("Syntax error!!")
        exit()
    elif charClass == DIGIT:
        addChar()
        getChar()
        getNonBlank()
        while charClass == DIGIT:
            addChar()
            getChar()
            getNonBlank()
        nextToken = INT_LIT
    elif charClass == UNKNOWN:
        lookup(nextChar)
        getChar()
    elif charClass == EOF:
        nextToken = EOF
        lexeme = 'EOF'
    
def addChar():
    global lexLen
    if lexLen <= BUF_MAX: 
        lexeme[lexLen] = nextChar
        lexLen += 1
        lexeme[lexLen] = '\0'
        
    else:
        print("lexeme is too long");
        exit()
    

def getChar():
    global input
    global checkCharCnt, charClass, nextChar
    nextChar = input[checkCharCnt]
    checkCharCnt += 1
    
    if nextChar != '\n':
        if nextChar.isalpha():
            charClass = LETTER
        elif nextChar.isdigit():
            charClass = DIGIT
        else:
            charClass = UNKNOWN
    else:
        
        charClass = EOF
    
def getNonBlank():
    global nextChar
    while (nextChar == ' ' or nextChar == '\t' ) and nextChar != '\n':
        getChar()
    
def lookup(ch):
    global nextToken, leftParenCnt, rightParenCnt
    if ch == '(':
        addChar();  nextToken = LEFT_PAREN; leftParenCnt += 1;
    elif ch == ')':
        addChar();  nextToken = RIGHT_PAREN; rightParenCnt += 1;
    elif ch == '+':
        addChar();  nextToken = ADD_OP;
    elif ch == '-':
        addChar();  nextToken = SUB_OP;
    elif ch == '*':
        addChar();  nextToken = MULT_OP;
    elif ch == '/':
        addChar();  nextToken = DIV_OP;
    else:
        addChar();  nextToken = EOF;

    return nextToken

def add(left, right):
    return left + right

def sub(left, right):
    return left - right

def mult(left, right):
    return left * right

def divd(left, right):
    return left / right
    
    
def main():
    global leftParenCnt, rightParenCnt, input, lexeme, cur_lexeme, nextChar, nextToken, token, charClass, lexLen, checkCharCnt

    while True:
        lexeme = [0 for i in range(BUF_MAX)]
        cur_lexeme = [0 for i in range(BUF_MAX)]
        nextChar = 0
        nextToken = token = 0
        charClass = 0
        lexLen = 0
        leftParenCnt = rightParenCnt = 0
        checkCharCnt = 0
        print('>>', end=' ', flush=True)
        input = list(sys.stdin.readline())
        if(input[0] == '\n'):
            continue
        getChar()
        lex()
        result = expr()
        if leftParenCnt != rightParenCnt:
            print("Syntax error!!");
            exit()
        if(result - int(result) == 0):
            print(int(result))
        else:
            print(result)    
        

if __name__ == '__main__':
    main()
    

File: assignment2/RD_parser.java
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class RD_parser {
    static final int LETTER = 0;
    static final int DIGIT = 1;
    static final int UNKNOWN = 99;
    static final int EOF = -1;

    static final int INT_LIT = 10;
    static final int ADD_OP = 21; // '+'
    static final int SUB_OP  = 22; // '-'
    static final int MULT_OP = 23; // '*'
    static final int DIV_OP = 24; // '/'
    static final int LEFT_PAREN = 25; // '('
    static final int RIGHT_PAREN = 26; // ')'
    static final int BUF_MAX = 1024;

    List<String> lexeme = new ArrayList<>();

    char nextChar = 0;
    int nextToken = 0;
    int charClass = 0;
    int lexLen = 0;
    int leftParenCnt = 0;
    int rightParenCnt = 0;
    int checkCharCnt = 0;
    String input;

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        while(true){
            RD_parser rd_parser = new RD_parser();
            System.out.print(">> ");
            String temp = sc.nextLine();
            rd_parser.input = temp + "\n";
            rd_parser.getChar();
            if(rd_parser.nextChar == '\n')
                continue;
            rd_parser.lex();
            double result = rd_parser.expr();
            if(rd_parser.leftParenCnt != rd_parser.rightParenCnt){
                System.out.println("Syntax error!!");
                System.exit(1);
            }
            if(result - (int)result == 0){
                System.out.println((int)result);
            }
            else {
                System.out.println(result);
            }
        }

    }
    double expr(){
        double left = term();
        double rslt = left;
        double right;
        int token;
        while(nextToken == ADD_OP || nextToken == SUB_OP){
            token = nextToken;
            lex();
            right = term();
            if(token == ADD_OP){
                rslt = add(left, right);
                left = rslt;
            }
            else if(token == SUB_OP){
                rslt = sub(left, right);
                left = rslt;
            }
        }
        return rslt;
    }

    double term(){
        double left = factor();
        double rslt = left;
        double right;
        int token;
        while(nextToken == MULT_OP || nextToken == DIV_OP){
            token = nextToken;
            lex();
            right = factor();
            if(token == MULT_OP){
                rslt = mult(left, right);
                left = rslt;
            }
            else if(token == DIV_OP){
                rslt = divd(left, right);
                left = rslt;
            }
        }
        return rslt;
    }

    double factor(){
        int ifUnary = 0;
        if(nextToken == SUB_OP){
            ifUnary = 1;
            lex();
        }
        if(nextToken == LEFT_PAREN){
            lex();
            double exprVal = expr();
            if(nextToken == RIGHT_PAREN){
                lex();
            }
            else{
                System.out.println("Syntax error!!");
                System.exit(1);
            }
            if(ifUnary == 1)
                return -exprVal;
            else
                return exprVal;
        }
        else if(nextToken == INT_LIT){
            if(ifUnary == 1)
                return -number();
            else
                return number();
        }
        else {
            System.out.println("Syntax error!!");
            System.exit(1);
        }
        return 0;
    }

    double number(){
        return digit();
    }

    double digit(){
        List<String> cur_lexeme = new ArrayList<>();
        cur_lexeme.addAll(lexeme);
        lex();
        return Float.parseFloat(String.join("", cur_lexeme));
    }

    void lex(){
        lexLen = 0;
        lexeme = new ArrayList<>();
        getNonBlank();
        switch (charClass){
            case LETTER:
                System.out.println("Syntax error!!");
                System.exit(1);
            case DIGIT:
                addChar();
                getChar();
                getNonBlank();
                while (charClass == DIGIT){
                    addChar();
                    getChar();
                    getNonBlank();
                }
                nextToken = INT_LIT;
                break;
            case UNKNOWN:
                lookup(nextChar);
                getChar();
                break;
            case EOF:
                nextToken = EOF;
                lexeme = new ArrayList<>();
                lexeme.add("EOF"+'\0');
                break;
            default:
                break;
        }
    }

    void addChar(){
        if(lexLen <= BUF_MAX){
            lexeme.add(lexLen++, nextChar+"");
            lexeme.add(lexLen, '\0'+"");
        }
        else{
            System.out.println("lexeme is too long");
            System.exit(1);
        }
    }

    void getChar(){
         nextChar = input.charAt(checkCharCnt++);
         if(nextChar != '\n'){
             if(Character.isLetter(nextChar))
                 charClass = LETTER;
             else if (Character.isDigit(nextChar))
                 charClass = DIGIT;
             else
                 charClass = UNKNOWN;
         }
         else
             charClass = EOF;
    }

    int lookup(char ch){
        switch (ch){
            case '(':	addChar();	nextToken = LEFT_PAREN;		leftParenCnt++;		break;
            case ')':	addChar();	nextToken = RIGHT_PAREN;	rightParenCnt++;	break;
            case '+':   addChar();	nextToken = ADD_OP;     	break;
            case '-':	addChar();	nextToken = SUB_OP;			break;
            case '*':	addChar();	nextToken = MULT_OP;		break;
            case '/':	addChar();  nextToken = DIV_OP;			break;
            case '\n':  addChar();  nextToken = EOF;            break;
            default:
                System.out.println("Syntax error!!");
                System.exit(1);
                break;
        }
        return nextToken;
    }

    void getNonBlank(){
        while((Character.isSpaceChar(nextChar) || nextChar == '\t') && nextChar != '\n')
            getChar();
    }

    double add(double left, double right){
        return left + right;
    }

    double sub(double left, double right){
        return left - right;
    }

    double mult(double left, double right){
        return left * right;
    }

    double divd(double left, double right){
        return left / right;
    }
}
